File: src/test/java/io/lsdconsulting/lsd/distributed/generator/integration/testapp/repository/TestRepository.java
package io.lsdconsulting.lsd.distributed.generator.integration.testapp.repository;

import com.mongodb.ConnectionString;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import de.flapdoodle.embed.mongo.MongodExecutable;
import de.flapdoodle.embed.mongo.MongodStarter;
import de.flapdoodle.embed.mongo.config.IMongodConfig;
import de.flapdoodle.embed.mongo.config.MongodConfigBuilder;
import de.flapdoodle.embed.mongo.config.Net;
import io.lsdconsulting.lsd.distributed.access.model.InterceptedInteraction;
import io.lsdconsulting.lsd.distributed.mongo.repository.codec.TypeCodec;
import io.lsdconsulting.lsd.distributed.mongo.repository.codec.ZonedDateTimeCodec;
import lombok.extern.slf4j.Slf4j;
import org.bson.Document;
import org.bson.codecs.configuration.CodecRegistries;
import org.bson.codecs.configuration.CodecRegistry;
import org.bson.codecs.pojo.PojoCodecProvider;
import org.bson.types.ObjectId;

import java.io.IOException;

import static com.mongodb.MongoClientSettings.builder;
import static com.mongodb.MongoClientSettings.getDefaultCodecRegistry;
import static de.flapdoodle.embed.mongo.distribution.Version.Main.PRODUCTION;
import static de.flapdoodle.embed.process.runtime.Network.localhostIsIPv6;
import static org.bson.codecs.configuration.CodecRegistries.fromProviders;
import static org.bson.codecs.configuration.CodecRegistries.fromRegistries;

@Slf4j
public class TestRepository {
    public static final String MONGODB_HOST = "localhost";
    public static final int MONGODB_PORT = 27017;

    private static final String DATABASE_NAME = "lsd";
    private static final String COLLECTION_NAME = "interceptedInteraction";

    private static MongoClient mongoClient;
    private static MongodExecutable mongodExecutable;

    private final CodecRegistry pojoCodecRegistry = fromRegistries(getDefaultCodecRegistry(), CodecRegistries.fromCodecs(new ZonedDateTimeCodec(), new TypeCodec()),
            fromProviders(PojoCodecProvider.builder().automatic(true).build()));

    public static void setupDatabase() {
        try {
            final IMongodConfig mongodConfig = new MongodConfigBuilder()
                    .version(PRODUCTION)
                    .net(new Net(MONGODB_HOST, MONGODB_PORT, localhostIsIPv6()))
                    .build();

            mongodExecutable = MongodStarter.getDefaultInstance().prepare(mongodConfig);
            mongodExecutable.start();


            mongoClient = MongoClients.create(builder()
                    .applyConnectionString(new ConnectionString("mongodb://" + MONGODB_HOST + ":" + MONGODB_PORT))
                    .retryWrites(true)
                    .build());
        } catch (final IOException e) {
            log.error(e.getMessage(), e);
        }
    }

    public static void tearDownDatabase() {
        mongoClient.close();
        mongodExecutable.stop();
    }

    public void save(InterceptedInteraction interceptedInteraction) {
        final MongoDatabase database = mongoClient.getDatabase(DATABASE_NAME);
        final MongoCollection<Document> collection = database.getCollection(COLLECTION_NAME).withCodecRegistry(pojoCodecRegistry);
        Document document = new Document("_id", new ObjectId());
        document.append("traceId", interceptedInteraction.getTraceId())
                .append("body", interceptedInteraction.getBody())
                .append("requestHeaders", interceptedInteraction.getRequestHeaders())
                .append("responseHeaders", interceptedInteraction.getResponseHeaders())
                .append("serviceName", interceptedInteraction.getServiceName())
                .append("target", interceptedInteraction.getTarget())
                .append("path", interceptedInteraction.getPath())
                .append("httpStatus", interceptedInteraction.getHttpStatus())
                .append("httpMethod", interceptedInteraction.getHttpMethod())
                .append("type", interceptedInteraction.getType())
                .append("profile", interceptedInteraction.getProfile())
                .append("elapsedTime", interceptedInteraction.getElapsedTime())
                .append("createdAt", interceptedInteraction.getCreatedAt());
        collection.insertOne(document);
    }
}

File: src/test/kotlin/io/lsdconsulting/lsd/distributed/generator/integration/testapp/repository/TestRepository.kt
package io.lsdconsulting.lsd.distributed.generator.integration.testapp.repository

import com.mongodb.ConnectionString
import com.mongodb.MongoClientSettings
import com.mongodb.client.MongoClient
import com.mongodb.client.MongoClients
import de.flapdoodle.embed.mongo.MongodExecutable
import de.flapdoodle.embed.mongo.MongodStarter
import de.flapdoodle.embed.mongo.config.MongodConfig
import de.flapdoodle.embed.mongo.config.Net
import de.flapdoodle.embed.mongo.distribution.Version
import de.flapdoodle.embed.process.runtime.Network
import io.lsdconsulting.lsd.distributed.connector.model.InterceptedInteraction
import io.lsdconsulting.lsd.distributed.generator.integration.testapp.config.log
import io.lsdconsulting.lsd.distributed.mongo.repository.codec.InteractionTypeCodec
import io.lsdconsulting.lsd.distributed.mongo.repository.codec.ZonedDateTimeCodec
import org.bson.Document
import org.bson.codecs.configuration.CodecRegistries
import org.bson.codecs.pojo.PojoCodecProvider
import org.bson.types.ObjectId
import java.io.IOException

class TestRepository {
    private val pojoCodecRegistry = CodecRegistries.fromRegistries(
        MongoClientSettings.getDefaultCodecRegistry(),
        CodecRegistries.fromCodecs(ZonedDateTimeCodec(), InteractionTypeCodec()),
        CodecRegistries.fromProviders(PojoCodecProvider.builder().automatic(true).build())
    )

    fun save(interceptedInteraction: InterceptedInteraction) {
        val database = mongoClient.getDatabase(DATABASE_NAME)
        val collection = database.getCollection(COLLECTION_NAME).withCodecRegistry(pojoCodecRegistry)
        val document = Document("_id", ObjectId())
        document.append("traceId", interceptedInteraction.traceId)
            .append("body", interceptedInteraction.body)
            .append("requestHeaders", interceptedInteraction.requestHeaders)
            .append("responseHeaders", interceptedInteraction.responseHeaders)
            .append("serviceName", interceptedInteraction.serviceName)
            .append("target", interceptedInteraction.target)
            .append("path", interceptedInteraction.path)
            .append("httpStatus", interceptedInteraction.httpStatus)
            .append("httpMethod", interceptedInteraction.httpMethod)
            .append("interactionType", interceptedInteraction.interactionType)
            .append("profile", interceptedInteraction.profile)
            .append("elapsedTime", interceptedInteraction.elapsedTime)
            .append("createdAt", interceptedInteraction.createdAt)
        collection.insertOne(document)
    }

    companion object {
        const val MONGODB_HOST = "localhost"
        const val MONGODB_PORT = 27017
        private const val DATABASE_NAME = "lsd"
        private const val COLLECTION_NAME = "interceptedInteraction"
        private lateinit var mongoClient: MongoClient
        private lateinit var mongodExecutable: MongodExecutable

        fun setupDatabase() {
            try {
                val mongodConfig: MongodConfig = MongodConfig.builder()
                    .version(Version.Main.V5_0)
                    .net(Net(MONGODB_HOST, MONGODB_PORT, Network.localhostIsIPv6()))
                    .build()
                mongodExecutable = MongodStarter.getDefaultInstance().prepare(mongodConfig)
                mongodExecutable.start()
                mongoClient = MongoClients.create(
                    MongoClientSettings.builder()
                        .applyConnectionString(ConnectionString("mongodb://$MONGODB_HOST:$MONGODB_PORT"))
                        .retryWrites(true)
                        .build()
                )
            } catch (e: IOException) {
                log().error(e.message, e)
            }
        }

        fun tearDownDatabase() {
            mongoClient.close()
            mongodExecutable.stop()
        }
    }
}